xc/device.ml

```diff
@@ -2227,6 +2231,79 @@ module Backend = struct
       let cmdline_of_info ~xs ~dm info restore domid =
         let common = Dm_Common.cmdline_of_info ~xs ~dm info restore domid ~domid_for_vnc:true in
 
+        let usb =
+        match info.Dm_Common.usb with
+        | Dm_Common.Disabled -> []
+        | Dm_Common.Enabled devices ->
+          let devs = devices |>
+            List.map (fun (x,y) -> ["-device"; sprintf "usb-%s,port=%d" x y]) |> List.concat in
+          "-usb" :: devs
+        in
+
+        let serial_device =
+          try
+            let xs_path = xs.Xs.read "/local/logconsole/@" in
+            let file =
+            if Stdext.Xstringext.String.has_substr xs_path "%d" then
+              Stdext.Xstringext.String.replace "%d" (string_of_int domid) xs_path
+            else
+              xs_path
+            in Some ("file:" ^ file)
+          with _ -> info.Dm_Common.serial
+        in
+
+        let list1 = ["piix3-ide-xen"; "piix3-usb-uhci"; "rtl8139"] in
+        let list2 = ["subvendor_id=0x5853"; "subsystem_id=0x0001"] in
+        let global = List.map (fun x -> List.map (fun y -> x^"."^y) list2) list1 |> List.concat in
+
+        let qmp = ["libxl"; "event"] |>
+          List.map (fun x -> ["-qmp"; sprintf "unix:/var/run/xen/qmp-%s-%d,server,nowait" x domid]) |>
+          List.concat in
+
+        let has_platform_device =
+          try
+            int_of_string (xs.Xs.read (sprintf "/local/domain/%d/vm-data/disable_pf" domid)) <> 1
+          with _ -> true
+        in
+        let xen_platform_device =
+          if has_platform_device then begin
+            let device_id =
+              try xs.Xs.read (sprintf "/local/domain/%d/platform/device_id" domid)
+              with _ -> "0001"
+            in
+            ["-device"; sprintf "xen-platform,addr=3,device-id=0x%s,revision=0x2,class-id=0x0100,subvendor_id=0x5853,subsystem_id=0x%s" device_id device_id]
+          end else []
+        in
+
+        let pv_device =
+          try
+            let has_device = xs.Xs.read (sprintf "/local/domain/%d/control/has-vendor-device" domid) in
+            if int_of_string has_device = 1 then ["-device"; "xen-pvdevice,device-id=0xc000"]
```

Comment:
Is conversion to int necessary?
```
if has_device = "1" ..
```

Reply:
The original script uses conversion, so I kept it, both here and above. I presume the reason is that conversion may be safer, for example, could the xenstore value be `001`? In this case string comparison will fail.